`make codegen` failed with many `rm: cannot remove [..]: Permission denied`

**Describe the bug**

`make codegen` failed with many `rm: cannot remove [..]: Permission denied` as below. See the Screenshot section to check out whole the logs.

```
++++++++++++++++++++++++++++++++
++++ GENERATING PROTO FILES ++++
++++++++++++++++++++++++++++++++
[..]
pkg/apis/proto/daemon/daemon.proto:22:1: warning: Import google/protobuf/empty.proto is unused.
rm: cannot remove '/tmp/tmp.C4WKeEqV6l/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/pkg/include/funcdata.h': Permission denied
rm: cannot remove '/tmp/tmp.C4WKeEqV6l/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/pkg/include/asm_ppc64x.h': Permission denied
rm: cannot remove '/tmp/tmp.C4WKeEqV6l/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/pkg/include/asm_riscv64.h': Permission denied
rm: cannot remove '/tmp/tmp.C4WKeEqV6l/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/pkg/include/asm_amd64.h': Permission denied
rm: cannot remove '/tmp/tmp.C4WKeEqV6l/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/pkg/include/textflag.h': Permission denied
[..]
rm: cannot remove '/tmp/tmp.C4WKeEqV6l/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/SECURITY.md': Permission denied
rm: cannot remove '/tmp/tmp.C4WKeEqV6l/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/bin/go': Permission denied
rm: cannot remove '/tmp/tmp.C4WKeEqV6l/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/bin/gofmt': Permission denied
rm: cannot remove '/tmp/tmp.C4WKeEqV6l/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/VERSION': Permission denied
rm: cannot remove '/tmp/tmp.C4WKeEqV6l/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/codereview.cfg': Permission denied
make: *** [Makefile:235: codegen] Error 1
```

I found that the cause was probably `./hack/generate-proto.sh` failed.

Also `make pre-push -B` (written in PR template) failed because it depends on `codegen`.

**To Reproduce**

Steps to reproduce the behavior:

1. Run any of the following commands:
    - `make pre-push -B`
    - `make codegen`
    - `./hack/generate-proto.sh`

**Expected behavior**

Any of them above passes.

**Screenshots**

* [01-make-pre-push-B-fails.log](https://github.com/user-attachments/files/19881191/01-make-pre-push-B-fails.log)
* [02-make-codegen-fails.log](https://github.com/user-attachments/files/19881195/02-make-codegen-fails.log)
* [03-hack-generate-proto-fails.log](https://github.com/user-attachments/files/19881255/03-hack-generate-proto-fails.log)

**Environment (please complete the following information):**

- make: 4.3
- go: 1.24.1
- protoc: 3.21.12
- Numaflow: commit d1bb56627695b0b08f1464d42b1a6e2675b2733e

**Additional context**

The following comment can help us.

https://github.com/golang/go/issues/35615#issuecomment-554301773

---


<!-- Issue Author: Don't delete this message to encourage other users to support your issue! -->

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We often sort issues this way to know what to prioritize.

>For quick help and support, join our [slack](https://numaproj.slack.com/join/shared_invite/zt-19svuv47m-YKHhsQ~~KK9mBv1E7pNzfg#/shared-invite/email) channel.

--- a/Makefile
+++ b/Makefile
@@ -24,6 +24,10 @@ GIT_BRANCH=$(shell git rev-parse --symbolic-full-name --verify --quiet --abbrev-
 GIT_TAG=$(shell if [[ -z "`git status --porcelain`" ]]; then git describe --exact-match --tags HEAD 2>/dev/null; fi)
 GIT_TREE_STATE=$(shell if [[ -z "`git status --porcelain`" ]]; then echo "clean" ; else echo "dirty"; fi)
 
+ifndef GOPATH
+GOPATH=$(shell go env GOPATH)
+endif
+
 DOCKER_PUSH?=false
 DOCKER_BUILD_ARGS?=
 IMAGE_NAMESPACE?=quay.io/numaproj
@@ -261,12 +265,12 @@ manifests: crds
 	kubectl kustomize config/extensions/webhook > config/validating-webhook-install.yaml
 
 $(GOPATH)/bin/golangci-lint:
-	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b `go env GOPATH`/bin v1.64.8
+	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(GOPATH)/bin v1.64.8
 
 .PHONY: lint
 lint: $(GOPATH)/bin/golangci-lint
 	go mod tidy
-	golangci-lint run --fix --verbose --concurrency 4 --timeout 5m --enable goimports
+	$(GOPATH)/bin/golangci-lint run --fix --verbose --concurrency 4 --timeout 5m --enable goimports
 
 .PHONY: start
 start: image

--- a/hack/generate-proto.sh
+++ b/hack/generate-proto.sh
@@ -10,6 +10,8 @@ header "generating proto files"
 ensure_vendor
 
 make_fake_paths
+# This overrides the existing trap
+trap 'cd "${FAKE_GOPATH}" && go clean -modcache && cd /tmp && rm -rf "${FAKE_GOPATH}"' EXIT
 export GOPATH="${FAKE_GOPATH}"
 export PATH="${GOPATH}/bin:${PATH}"
 cd "${FAKE_REPOPATH}"

